nexus/src/app/instance.rs

```diff
@@ -1183,6 +1183,28 @@ impl super::Nexus {
         let ssh_keys: Vec<String> =
             ssh_keys.map(|ssh_key| ssh_key.public_key).collect();
 
+        // Construct instance metadata used to track its statistics.
+        //
+        // This requires another fetch on the silo and project, to extract their
+        // IDs and names.
+        let (.., db_project) = self
+            .project_lookup(
+                opctx,
+                params::ProjectSelector {
+                    project: NameOrId::Id(db_instance.project_id),
+                },
+            )?
+            .fetch()
+            .await?;
+        let (_, db_silo) = self.current_silo_lookup(opctx)?.fetch().await?;
+        let metadata = sled_agent_client::types::InstanceMetadata {
+            silo_id: db_silo.id(),
+            silo_name: db_silo.name().to_string(),
+            project_id: db_project.id(),
+            project_name: db_project.name().to_string(),
```

Comment:
I think this is a reasonable place to draw the line. It also has some drawbacks. It means Nexus needs to understand more of the query structure, so that it can do things like translate a silo name into a silo ID. If there's some complicated query passed in, that might be a lot of work. That may also be unavoidable. I'm still not sure how auth will intersect with querying, but Nexus might already need to know what components of a query need to be checked for authz. E.g., Nexus would already need to be able to understand the query enough to know that something like `silo_name == "my-silo"` involves some authz checks.

As an alternative, we could imagine including the names in the metrics, and then require that updates to those names be propagated to the relevant leaves. For example, if a user updates a silo name, then we might have a task in Nexus that sends the new name(s) to all the sled-agents hosting instances in that silo. The sled-agents could then forward those to the relevant Propolis servers as well, which could update the instance metadata they publish to oximeter.

That has drawbacks too. The data may include names that do not exist as far as the Nexus API is concerned, which might be confusing or present authz problems of its own. (How do you check that a user is allowed to see metrics from a silo, if that silo no longer has that name? The name-to-ID lookup will fail.) We might be able to just obviate a lot of these by requiring that queries only refer to immutable IDs, rather than updateable names.

I'm not sure the right answer here.

Reply:
Sounds good, will update soon.

Reply:
Removed in 5ac187ddd